DarkHorseCorder/apptractive-fullstack-technical-test
# Development user story 001 006

### TEST001 - Set up Xero App

1. When logged in to Xero, Set app name as 'Fullstack Test' at https://developer.xero.com/app/manage/
2. Select integration type as "Web app"
3. Enter application url as https://localhost:4200
4. Enter the redirect URI as https://localhost:4200/xero-redirect (it will be used in oauth2 to exchange code for access token) 
5. Once app is created, generate the client ID and Client Secret for application access

### TEST002 - Connect to Xero API

1. Insert the generated Xero client ID and Client Secret into the project to allow connecting to the Xero API (search `xeroClientId` and `xeroClientSecret` in the project)
2. Ensure client Secret is not exposed in the code

### TEST003 - Set up AWS CDK

1. Authenticate to your AWS account
2. Set up AWS CDK using `npm install -g aws-cdk` or `yarn global add aws-cdk` if not already installed
3. If you have never used CDK in your AWS account, you will need to bootstrap it first `npx nx bootstrap backend --profile enterprofilename`
4. First deploy the lambda layer stack using `npx nx deploy backend FULLayerStack --profile enterprofilename`
5. Deploy all stacks in the backend using `npx nx deploy backend --all --require-approval=never --outputs-file output.json --profile enterprofilename`
6. Copy output file to react-app application by running `yarn run output` 

### TEST004 - Authenticated user - Connect Xero button

1. Once a user signs in, they should see the dashboard
2. On the dashboard there is a button "Connect Xero"
3. Clicking "Connect Xero" should open a new window to Xero oAuth 2.0
4. Once the user connects to Xero, they should be redirected back to the application (`/xero-redirect` route)

### TEST005 - Xero oauth2 - Exchange authorisation code for token set

1. Backend uses the redirect URL to exchange the authorisation code for a token set
2. The token set is stored for my user in the database for subsequent Xero API requests
3. Once successfully authenticated, in the web application the user should see a message that says "Xero Connected"
4. Once successfully authenticated, in the web application the user should see a button "Go To Dashboard"
5. Clicking "Go to Dashboard", user should navigate to the dashboard page

### TEST006 - Xero API Invoices

1. Retrieve the token set from the user's profile to call the Xero API 
2. If the token set is expired, the backend should refresh the token set before calling the Xero API
3. In the web application, the user should see a list of their invoices:
   - For each invoice, it should show the invoice number, status, amount paid, Tax total and total 
4. User should be able to toggle between paginated results
5. User should be able to filter invoices by status
